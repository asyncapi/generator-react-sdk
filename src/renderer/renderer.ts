function normalizeProps(props: any = {}) {
  const children = props.children;
  let newChildren = render(children);

  return { 
    ...props, 
    children: newChildren,
    $$children: children,
  }
}

function createElement(element: any): any {
  if (typeof element.type === 'string') {
    // HTML (also not standard) tags case
    throw new Error("HTML tags is not supported yet!");
  } else if (typeof element.type === 'symbol') {
    // internal React types like Fragments, Portal etc. We skip them.
    return render(element.props.children);
  } else if (typeof element.type === 'function') {
    const { type, props } = element;
    const prototype = type.prototype;

    // Class component case
    if (prototype && typeof prototype.isReactComponent === "object") {
      const clazzComp = new type(normalizeProps(props));
      return createElement(clazzComp.render());
    } 
    // Function component case
    return createElement(type(normalizeProps(props)));
  }
  return element || "";
}

export function render(node: React.ReactNode) {
  let content = "";
  const typeOf = typeof node;
  if (typeOf === 'string') {
    content += node;
  } else if (Array.isArray(node)) {
    content += node.map(child => {
      const childValue = createElement(child);
      return render(childValue);
    }).join("");
  } else {
    content += createElement(node);
  }
  return content || "";
}

export function renderTemplate(element: any = {}) {
  const { type, props = {} } = element;

  if (typeof type !== "function" && type.name !== "File") {
    throw new Error("File is required as first node in template!");
  }
  if (props.name === undefined) {
    throw new Error("`name` prop in File component is required!");
  }

  return {
    metadata: {
      name: props.name,
      permissions: props.permissions,
    },
    content: render(props.children),
  }
}
