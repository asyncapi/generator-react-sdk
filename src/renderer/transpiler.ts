

import Path from 'path';
import fs from 'fs';
import { promisify } from 'util';
const ROOT_DIR = Path.resolve(__dirname, '../..');
import { rollup, RollupBuild } from 'rollup';
import commonjs from '@rollup/plugin-commonjs';
import babel from '@rollup/plugin-babel';
const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);

/**
 * Options for transpiling files.
 */
export class TranspileFilesOptions {
    /**
     * Should all files in a directory including those in subdirectories be included
     */
    recursive: boolean = false
}

/**
 * Function which finds all the files in directories
 * 
 * @param dir directory to find files in.
 * @param includeSubDirs should the function iterate through subdirectories to search for files?
 */
async function getFilesInDir(dir: string, includeSubDirs: boolean) {
    const subdirs = await readdir(dir);
    const files: string[][] = await Promise.all(subdirs.map(async (subdir: string) => {
        const res = Path.resolve(dir, subdir);
        const isDir = (await stat(res)).isDirectory(); 
        if(!includeSubDirs && isDir) return [];
        return isDir ? getFilesInDir(res, includeSubDirs) : [res];
    }));
    return files.reduce((a, f) => a.concat(f), []);
}


/**
 * Transpile files in directory and write it to a directory, if no errors are thrown it completed successfully.
 * 
 * @param directory to transpile.
 * @param outputDir to write the transpiled files to.
 * @param options any extra options that should be passed.
 */
export async function transpileFiles(directory: string, outputDir: string, options?: TranspileFilesOptions) {
    const parsedFile = Path.parse(directory)
    const dir = parsedFile.dir
    const files = await getFilesInDir(dir, options !== undefined ? options.recursive : false);
    if(files.length == 0) throw new Error("No files to transpile?")

    let bundles: RollupBuild;
    try {
        bundles = await rollup({
            input: files,
            plugins: [
                babel({
                    cwd: ROOT_DIR,
                    presets: [
                        "@babel/preset-env",
                        "@babel/preset-react"
                    ],
                }),
                commonjs(),
            ]
        })
    } catch (e) {
        const transpileError = new Error("Could not transpile one or more files in the directory...");
        transpileError.stack = e;
        throw transpileError;
    }
    try {
        await bundles.write({
            format: "commonjs",
            sourcemap: true,
            dir: outputDir
        })
    } catch (e) {
        const transpileSaveError = new Error("Could not save one or more transpiled files in the directory...");
        transpileSaveError.stack = e;
        throw transpileSaveError;
    }
}