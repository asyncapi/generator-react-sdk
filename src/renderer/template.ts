import React from "react";
import { render } from "./renderer";
import Path from 'path';
import Fs from 'fs';
import { transpileFiles } from "./transpiler";

/**
 * Imports a given file and return the imported component
 * 
 * @param filepath to import
 */
function importComponent(filepath:string): Promise<React.ReactElement> {
  return new Promise((resolve) => {
    import(filepath).then(component => { resolve(component.default()) })
  })
}


    
/**
 * render a file with react. This function automatically transforms jsx to js before importing the component.
 * 
 * @param filepath file to render
 */
export async function renderTemplate(filepath: string, context: any, debug: boolean = false) {
  const parsedFile = Path.parse(filepath)
  const dir = parsedFile.dir
  const outputDir = Path.resolve(dir, '../__transpiled');
  //compile the template dir first
  await transpileFiles(filepath, outputDir, {recursive: true});

  
  const inputFile = Path.resolve(outputDir, parsedFile.base);
  const { type, props = {} } = await importComponent(inputFile);

  if (typeof type !== "function" || type.name !== "File") {
    throw new Error("File is required as first node in template!");
  }
  
  if (debug !== true) Fs.rmdirSync(outputDir, { recursive: true });


  return {
    metadata: {
      name: props.name,
      permissions: props.permissions,
    },
    content: render(props.children),
  }
}
