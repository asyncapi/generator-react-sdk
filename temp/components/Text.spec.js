import React from 'react';
import {describe, expect, test} from '@jest/globals';
import renderer from 'react-test-renderer';
import {Text, IndentationTypes} from '../../index';

describe('<Text />', () => {
  test('Should always render text as is', () => {
    const defaultProps = {};
    const wrapper = renderer.create(<Text {...defaultProps}>Test</Text>);
    expect(wrapper).toMatchSnapshot('text_should_render');
  });
  test('Should always render indentation with spaces', () => {
    const defaultProps = {
      size: '4',
      type: IndentationTypes.SPACES
    };
    const wrapper = renderer.create(<Text {...defaultProps}>Test</Text>);
    expect(wrapper).toMatchSnapshot('text_should_render_space_indentation');
  });
  test('Should always render indentation with tabs', () => {
    const defaultProps = {
      size: '4',
      type: IndentationTypes.TABS
    };
    const wrapper = renderer.create(<Text {...defaultProps}>Test</Text>);
    expect(wrapper).toMatchSnapshot('text_should_render_tab_indentation');
  });
  test('Should be able to handle wrong props values', () => {
    const defaultProps = {
      size: 'test',
      type: 'someweirdtype'
    };
    expect(() => renderer.create(<Text {...defaultProps}>Test</Text>)).toThrowErrorMatchingSnapshot('text_with_wrong_props_should_throw_error');
  });
});