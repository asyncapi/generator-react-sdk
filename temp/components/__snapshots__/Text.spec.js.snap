// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> Should always render indentation with spaces: text_should_render_space_indentation 1`] = `
"Test
"
`;

exports[`<Text /> Should always render indentation with tabs: text_should_render_tab_indentation 1`] = `
"Test
"
`;

exports[`<Text /> Should always render text as is: text_should_render 1`] = `
"Test
"
`;
